src/Orleans.Core/GrainDirectory/IGrainLocator.cs

```diff
@@ -0,0 +1,24 @@
+using System;
+using System.Collections.Generic;
+using System.Text;
+using System.Threading.Tasks;
+using Orleans.Runtime;
+
+namespace Orleans.GrainDirectory
+{
+    /// <summary>
+    /// Used to locate Grain activation in the cluster
+    /// </summary>
+    interface IGrainLocator
+    {
+        Task<ActivationAddress> Register(ActivationAddress address);
+
+        Task Unregister(ActivationAddress address, UnregistrationCause cause);
+
+        Task UnregisterMany(List<ActivationAddress> addresses, UnregistrationCause cause);
+
+        Task<List<ActivationAddress>> Lookup(GrainId grainId);
+
+        bool TryLocalLookup(GrainId grainId, out List<ActivationAddress> addresses);
```

Comment:
Can this and Lookup not be merged into one `ValueTask<List<ActivationAddress>>` method?

Reply:
Not "easily" at least, since in `Catalog` we have two explicit different method: `Catalog.FastLookup` and `Catalog.FullLookup`. The synchronous method is used by the `Dispatcher` when trying to address a message synchronously.

We could definitely change that also to simplify the flow.